plotutils/info/plotfont.1
.TH PLOTFONT 1 "Jun 2000" "FSF" "GNU Plotting Utilities"
.SH NAME
plotfont \- produce character maps of fonts supported by the plotting utilities
.\" Not all man macros define SB
.de SB
\&\fB\s-1\&\\$1 \\$2\s0\fR
..
.SH SYNOPSIS
.B plotfont
[
.I options 
]
.I fonts
.SH DESCRIPTION
.LP
.B plotfont
produces a character map for any font that is supported by the plotting
utilities, which include
.BR graph (1),
.BR plot (1),
.BR pic2plot (1),
.BR tek2plot (1),
and the GNU libplot 2-D graphics export library (see
.BR plot (3)).
Which fonts are supported depends on the output format,
which is specified by the 
.BR \-T " option."
A listing of the fonts available in any specified output format may be
obtained with the
.B \-\-help\-fonts
option (see below).
.LP
The character map, or maps, will be written to standard output in the
specified format.
For example, the Times-Roman font is available when producing Postscript
output.
The command
.B plotfont \-T ps Times\-Roman > charmap.ps
will yield a character map of the Times-Roman font, in a Postscript format
that can be viewed or edited with the
.BR idraw (1)
drawing editor.
The Times-Roman font is also available when producing Fig output, which
can be viewed or edited with the 
.BR xfig (1)
drawing editor.
The command
.B plotfont \-T fig Times\-Roman > charmap.fig
will yield the same character map, but in Fig format rather than
in Postscript format.
.LP
As another example, the Univers font is available when producing PCL 5
output.
The command
.B plotfont \-T pcl Univers > charmap.pcl
will produce a character map of the Univers font, in PCL 5 format.
.LP
When producing output for the X Window System, i.e., for a popped-up
window, any scalable X Window System font that has an
XLFD (i.e., X Logical Font Description) name is supported.
For example, the command
.B plotfont \-T X utopia\-medium\-r\-normal
will pop up a window, and draw a character map of the Utopia-Regular font.
"utopia-medium-r-normal" is a truncated version of the Utopia-Regular 
font's XLFD name.
The Utopia-Regular font is available on most X Window System displays.
.SH OPTIONS
.SS General Options
.TP
.BI \-T " type"
.br
.ns
.TP
.BI \-\-output\-format " type"
Select 
.I type
as the output format.
It may be "X", "png", "pnm", "gif", "svg", "ai", "ps", "cgm", "fig",
"pcl", "hpgl", "regis", "tek", or "meta" (the default).
These refer respectively
to the X Window System, 
PNG (Portable Network Graphics) format,
portable anymap format (PBM/PGM/PPM), 
a pseudo-GIF format that
does not use LZW encoding,
the new XML-based Scalable Vector Graphics format,
the format used by Adobe Illustrator, Postscript or
Encapsulated Postscript (EPS) that can be edited with
.BR idraw (1),
CGM format (by default, confirming to the WebCGM profile),
the format used by the 
.BR xfig (1) 
drawing editor, the Hewlett\-Packard PCL 5 printer
language, the Hewlett\-Packard Graphics Language, 
ReGIS graphics format (which can be displayed 
by the
.BR dxterm (1)
terminal emulator or by a VT330 or VT340 terminal),
Tektronix format (which can be displayed by the
.BR xterm (1)
terminal emulator),
and device-independent GNU metafile format itself.
Unless \fItype\fP\^ is "X", an output file is produced and written
to standard output.
.IP ""
Files in PNG, PNM, pseudo-GIF, AI, or Fig format contain only a single page
of graphics.
So if the
.B \-T png
option, the 
.B \-T pnm
option, the 
.B \-T gif
option, the 
.B \-T ai
option, or the 
.B \-T fig
option is used, the output file will contain a character map for only the
first-specified font.
.IP ""
A listing of the fonts available in any specified output format may be
obtained with the
.B \-\-help\-fonts
option (see below).
If a requested font is unavailable, a default font will be substituted.
The default font
is "Helvetica" for "X", "svg", "ai", "ps", "cgm", and "fig",
"Univers" for "pcl",
and "HersheySerif" for "png", "pnm", "gif", "hpgl", "regis", "tek", and "meta".
.TP
.B \-1
.br
.ns
.TP
.B \-\-lower\-half
Generate a character map for the lower half of each specified font.
This is the default.
.TP
.B \-2
.br
.ns
.TP
.B \-\-upper\-half
Generate a character map for the upper half of each specified font.
.TP
.B \-o
.br
.ns
.TP
.B \-\-octal
Number the characters in octal rather than in decimal (the default).
.TP
.B \-x
.br
.ns
.TP
.B \-\-hexadecimal
Number the characters in hexadecimal rather than in decimal (the default).
.TP
.B \-\-box
Surround each character with a box, showing its extent to left and right.
The default is not to do this.
.TP
.BI \-j " row"
.br
.ns
.TP
.BI \-\-jis\-row " row"
Generate a character map for row 
.I row
of a Japanese font arranged according to JIS [Japanese Industrial Standard]
X0208.
The only such font currently available is the HersheyEUC [Extended
Unix Code] font.
If used, this option overrides the 
.B \-1
and 
.B \-2
options.
The valid rows are 1.\|.\|.94.
In the JIS X0208 standard, Roman characters are located in row 3, and
Japanese syllabic characters (Hiragana and Katakana) are located in rows 4
and 5.
Greek and Cyrillic characters are located in rows 6 and 7.
Japanese ideographic characters (Kanji) are located in rows 16.\|.\|.84.
.TP
.BI \-\-bg\-color " name"
Set the color used for the background to be
.IR name .
This is relevant only to 
.BR "plotfont \-T X" ,
.BR "plotfont \-T png" ,
.BR "plotfont \-T pnm" ,
.BR "plotfont \-T gif" ,
.BR "plotfont \-T svg" ,
.BR "plotfont \-T cgm" ,
and
.BR "plotfont \-T regis" .
An unrecognized name sets the color to the default, which is "white".
The environment variable
.SB BG_COLOR
can equally well be used to specify the background color.
.IP ""
If the 
.B \-T png
or
.B \-T gif
option is used, a transparent PNG file
or a transparent pseudo-GIF, respectively, may be produced by
setting the 
.SB TRANSPARENT_COLOR
environment variable to the name of the background color.
If the 
.B \-T svg
or
.B \-T cgm
option is used, an output file without a background may be produced
by setting the background color to "none".
.TP
.BI \-\-bitmap\-size " bitmap_size"
Set the size of the graphics display in which the character map(s) 
will be drawn,
in terms of pixels, to be
.IR bitmap_size .
The default is "570x570".
This is relevant only to 
.BR "plotfont \-T X" , 
.BR "plotfont \-T png" , 
.BR "plotfont \-T pnm" , 
and
.BR "plotfont \-T gif" ,
all of which produce bitmaps.
If you choose a rectangular (non-square) window size, the fonts in the
character map(s) will be scaled anisotropically, i.e., by different factors
in the horizontal and vertical directions.
For
.BR "plotfont \-T X" , 
this requires an X11R6 display.
Any font that cannot be scaled in this way will be replaced by a default
scalable font, such as the vector font "HersheySerif".
.IP ""
The environment variable 
.SB BITMAPSIZE
can equally well be used to specify the window size.
For backward compatibility, the X resource 
.B Xplot.geometry
may be used instead.
.TP
.BI \-\-emulate\-color " option"
If 
.I option
is 
.IR yes ,
replace each color in
the output by an appropriate shade of gray.
This is seldom useful,
except when using 
.B plotfont \-T pcl
to prepare output for a PCL 5 device.
(Many monochrome PCL 5 devices, such as monochrome
LaserJets, do a poor job of emulating color on their own.)
You may equally well request color emulation by setting the environment
variable
.SB EMULATE_COLOR
to "yes".
.TP
.BI \-\-numbering\-font " name"
Set the font used for the numbering of the characters in the character
map(s) to be
.IR name ,
rather than the default.
.TP
.BI \-\-page\-size " pagesize"
Set the size of size of the page on which the character map(s) will be
positioned.
This is relevant only to
.BR "plotfont \-T svg" ,
.BR "plotfont \-T ai" ,
.BR "plotfont \-T ps" ,
.BR "plotfont \-T cgm" ,
.BR "plotfont \-T fig" ,
.BR "plotfont \-T pcl" ,
and
.BR "plotfont \-T hpgl" .
The default is "letter", which means an 8.5 inch by 11 inch page.
Any ISO page size in the range "a0".\|.\|."a4" or ANSI page size in the
range "a".\|.\|."e" may be specified ("letter" is an alias for "a" and
"tabloid" is an alias for "b").
"legal" and "ledger" are recognized page sizes also.
The environment variable
.SB PAGESIZE 
can equally well be used to specify the page size.
.IP ""
The graphics display in which each character map is drawn will be a square
region that would occupy nearly the full width of the specified page.
An alternative size for the graphics display can be specified.
For example, the page size could be specified as
"letter,xsize=4in,ysize=6in", or "a4,xsize=5.0cm,ysize=100mm".
For all of the above except
.BR "plotfont \-T hpgl" , 
the graphics display will, by default, be centered on the page.
For all of the above except
.B "plotfont \-T svg"
and
.BR "plotfont \-T cgm" ,
the graphics display may be repositioned manually, by specifying the
location of its lower left corner, relative to the lower left corner of the
page.
For example, the page size could be specified as
"letter,xorigin=2in,yorigin=3in", or "a4,xorigin=0.5cm,yorigin=0.5cm".
It is also possible to specify an offset vector.
For example, the page size could be specified as "letter,xoffset=1in",
or "letter,xoffset=1in,yoffset=1.2in", or "a4,yoffset=\-1cm".

CHANGELOG
1.7.3 (not yet released)
========================

Features
--------
* Plots now show inline in jupyter notebooks.
* Added docs/examples for jupyter notebooks.
* Can set the default background color for plots using `bgcolor = ` in the 
  `[default]` section of the config.  This default effects plots shown on the 
  screen and non-antialiased plots written with write_img.  The default
  can specifically be overridden in the `[screen]` section and the new section
  configuring non-anti-aliased plots `[image_noaa]`.
* New config section for anti-aliased images `[image]` where currently you can
  set the resolution, which defaults to `dpi = 100`. This can be overridden
  in call to `write(..., dpi=)`
* New config section for non-anti-aliased images `[image_noaa]` where currently you can
  set the default `width =`, `height=`, `bgcolor=` and `color = `.  Default `width,height`
  is 640.  The colors default to those set in the `[default]` section.
* The `write_img` function can now be called with just the filename, since
  the default `width,height` can now be set in the `[image_noaa]` section.
  The other arguments can now be sent with keywords (old style still supported).

Bug Fixes
----------

* make ScreenRender close itself after leaving the context
  manager (or being cleaned up). This prevents a seg fault that could occur
  when the X11 device was open and an error occured, and the device was not
  then closed.  This was basically a mistake on my part thinking
  that close would close the window, so I had not deleted it.

Removed Features
----------------

* removed the `persistent = yes` "feature" for X11 windows.  This never
  worked.  It may be re-implemented in the future.


1.7.2 (14 Mar 2017)
===================

Features
--------
* Added `+=` syntax for adding components to a `PlotKey`.
* Added `+=` syntax for adding components to containers.
* Added `Polygon` component.

Bug Fixes
---------
* Added /usr/X11 to search path for linking plotutils. (Thank you @smaret!)

Code Changes
------------
* Updated examples to use new `+=` syntax.
* Updated code to PEP8 with `autopep8`.
* Updated travis-ci to ship to pypi on tags.
* Added `MANIFEST.in` file to include `README.rst` in the package.

1.7.0 (08 Mar 2017)
===================

Features
--------
* Increased default screen size to 640x640.
* Changed default font to `HersheySerif`.
* Changed default to non-persistent windows.
* Added new write functions for anti-aliased images.
* Added smooth keyword to histogram.
* Added option to change panel ratios in `FramedArray` using
  `row_fractions` and `col_fractions` keywords.
* Visibility of panels in `FramedArray` can be controlled
  with the visibility attribute, `arr[i,j].visible=False`.
* Added `\odot` to latex conversion, for sun symbol.
* New `func.plot` function with more functionality.
* Major classes and functions documented. This facilitates
  checking docs from an interactive session, e.g. `help()` or `?` in ipython.
* New documentation on GitHub wiki.

Bug Fixes
---------
* Fixed plotting of vertical and horizontal lines on log axes.
* Fixed bug in bounding box unions computing minimum.
* Added `try: finally:` blocks around compose operations. Fixes segfault
  when an exception is thrown while composing and more plotting is attempted.
* Fixed bug in `DataArc`/`PlotArc`.

Code Changes
------------
* Added Travis CI for testing.
* Converted to new style classes.
* Converted to extended calling syntax.
* Moved to spaces instead of tabs.

Removed Code
------------
* Removed old C wrapper for GNU plotutils.
* Removed old Makefile build system.
* Removed matlab compatible interface.
* Removed read_* I/O code from `func.py`.

1.6.7 (02 May 2012)
===================

* Fix segfault when libplot isn't built with X11 (reported by Daniel Ericsson
  & Sebastien Maret).

* Move to github.

1.6.6 (27 Nov 2008)
===================

* Fixed issue that caused all images to be 570x570 on some systems, due to a
  change in the python2.5 C API.

* Improved setup.py and Makefiles.

* Contours can now be used with PlotKey.

* Added Labels component. Similar to Points, but with text labels instead
  of symbols.

1.6.5 (20 Mar 2007)
===================

* Replaced Numeric with numpy.

1.6.4 (08 Mar 2004)
===================

* Phil Kromer contributed the new ColoredPoints and Density components.
  See example9.py for details.

* Todd Fox contributed Makefiles for building with MS Visual C++.

* Can now specify the width/height of postscript output, by passing (e.g.)

plotutils/plotfont/plotfont.c
/* This file is part of the GNU plotutils package.  Copyright (C) 1995,
   1996, 1997, 1998, 1999, 2000, 2005, 2008, 2009, Free Software
   Foundation, Inc.

   The GNU plotutils package is free software.  You may redistribute it
   and/or modify it under the terms of the GNU General Public License as
   published by the Free Software foundation; either version 2, or (at your
   option) any later version.

   The GNU plotutils package is distributed in the hope that it will be
   useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with the GNU plotutils package; see the file COPYING.  If not, write to
   the Free Software Foundation, Inc., 51 Franklin St., Fifth Floor,
   Boston, MA 02110-1301, USA. */

/* This file is the driving routine for the GNU `plotfont' program.  It
   includes code to plot all characters in a specified font. */

#include "sys-defines.h"
#include "libcommon.h"
#include "getopt.h"
#include "fontlist.h"
#include "plot.h"

const char *progname = "plotfont"; /* name of this program */
const char *written = "Written by Robert S. Maier.";
const char *copyright = "Copyright (C) 2009 Free Software Foundation, Inc.";

const char *usage_appendage = " FONT...\n";

enum radix { DECIMAL, OCTAL, HEXADECIMAL };

/* options */

#define	ARG_NONE	0
#define	ARG_REQUIRED	1
#define	ARG_OPTIONAL	2

const char *optstring = "12oxOj:J:F:T:";

struct option long_options[] = 
{
  /* The most important option ("--display-type" is an obsolete variant) */
  { "output-format",	ARG_REQUIRED,	NULL, 'T'},
  { "display-type",	ARG_REQUIRED,	NULL, 'T' << 8 }, /* hidden */
  /* Other frequently used options */
  { "box",		ARG_NONE,	NULL, 'b' << 8 },
  { "octal",		ARG_NONE,	NULL, 'o'},
  { "hexadecimal",	ARG_NONE,	NULL, 'x'},
  { "jis-page",		ARG_REQUIRED,	NULL, 'J'},
  { "jis-row",		ARG_REQUIRED,	NULL, 'j'},
  { "lower-half",	ARG_NONE,	NULL, '1'},
  { "upper-half",	ARG_NONE,	NULL, '2'},
  { "font-name",	ARG_REQUIRED,	NULL, 'F' }, /* hidden */
  /* Long options with no equivalent short option alias */
  { "numbering-font-name", ARG_REQUIRED, NULL, 'N' << 8 },
  { "title-font-name",	ARG_REQUIRED,	NULL, 'Z' << 8 },
  { "pen-color",	ARG_REQUIRED,	NULL, 'C' << 8 },
  { "bg-color",		ARG_REQUIRED,	NULL, 'q' << 8 },
  { "bitmap-size",	ARG_REQUIRED,	NULL, 'B' << 8 },
  { "emulate-color",	ARG_REQUIRED,	NULL, 'e' << 8 },  
  { "page-size",	ARG_REQUIRED,	NULL, 'P' << 8 },
  { "rotation",		ARG_REQUIRED,	NULL, 'r' << 8 },
  /* Options relevant only to raw plotfont (refers to metafile output) */
  { "portable-output",	ARG_NONE,	NULL, 'O' },
  /* Documentation options */
  { "help-fonts",	ARG_NONE,	NULL, 'f' << 8 },
  { "list-fonts",	ARG_NONE,	NULL, 'l' << 8 },
  { "version",		ARG_NONE,	NULL, 'V' << 8 },
  { "help",		ARG_NONE,	NULL, 'h' << 8 },
  { NULL,		0,		NULL,  0}
};
    
/* null-terminated list of options, such as obsolete-but-still-maintained
   options or undocumented options, which we don't show to the user */
const int hidden_options[] = { (int)'J', (int)'F', (int)('T' << 8), 0 };

/* forward references */
bool do_font (plPlotter *plotter, const char *name, bool upper_half, char *pen_color_name, char *numbering_font_name, char *title_font_name, bool bearings, enum radix base, int jis_page, bool do_jis_page);
void write_three_bytes (int charnum, char *numbuf, int radix);
void write_two_bytes (int charnum, char *numbuf, int radix);


int
main (int argc, char *argv[])
{
  plPlotter *plotter;
  plPlotterParams *plotter_params;
  bool bearings = false;	/* show sidebearings on characters? */
  bool do_jis_page = false;	/* show page of HersheyEUC in JIS encoding? */
  bool do_list_fonts = false;	/* show a list of fonts? */
  bool show_fonts = false;	/* supply help on fonts? */
  bool show_usage = false;	/* show usage message? */
  bool show_version = false;	/* show version message? */
  bool upper_half = false;	/* upper half of font, not lower? */
  char *output_format = (char *)"meta"; /* default libplot output format */
  char *numbering_font_name = NULL; /* numbering font name, NULL -> default */
  char *option_font_name = NULL; /* allows user to use -F */
  char *pen_color = NULL;	/* initial pen color, can be spec'd by user */
  char *title_font_name = NULL;	/* title font name, NULL -> current font */
  enum radix base = DECIMAL;
  int errcnt = 0;		/* errors encountered */
  int jis_page = 33;		/* page of HersheyEUC in JIS encoding */
  int opt_index;		/* long option index */
  int option;			/* option character */
  int retval;			/* return value */

  plotter_params = pl_newplparams ();
  while ((option = getopt_long (argc, argv, optstring, long_options, &opt_index)) != EOF)
    {
      if (option == 0)
	option = long_options[opt_index].val;
      
      switch (option) 
	{
	case 'T':		/* Output format, ARG REQUIRED      */
	case 'T' << 8:
	  output_format = (char *)xmalloc (strlen (optarg) + 1);
	  strcpy (output_format, optarg);
	  break;
	case 'O':		/* Ascii output */
	  pl_setplparam (plotter_params, "META_PORTABLE", (char *)"yes");
	  break;
	case '1':		/* Lower half */
	  upper_half = false;
	  break;
	case '2':		/* Upper half */
	  upper_half = true;
	  break;
	case 'o':		/* Octal */
	  base = OCTAL;
	  break;
	case 'x':		/* Hexadecimal */
	  base = HEXADECIMAL;
	  break;
	case 'F':		/* set the initial font */
	  option_font_name = (char *)xmalloc (strlen (optarg) + 1);
	  strcpy (option_font_name, optarg);
	  break;
	case 'e' << 8:		/* Emulate color via grayscale */
	  pl_setplparam (plotter_params, "EMULATE_COLOR", (char *)optarg);
	  break;
	case 'N' << 8:		/* Numbering Font name, ARG REQUIRED */
	  numbering_font_name = xstrdup (optarg);
	  break;
	case 'Z' << 8:		/* Title Font name, ARG REQUIRED */
	  title_font_name = xstrdup (optarg);
	  break;
	case 'C' << 8:		/* set the initial pen color */
	  pen_color = (char *)xmalloc (strlen (optarg) + 1);
	  strcpy (pen_color, optarg);
	  break;
	case 'q' << 8:		/* set the initial background color */
	  pl_setplparam (plotter_params, "BG_COLOR", (void *)optarg);
	  break;
	case 'B' << 8:		/* Bitmap size */
	  pl_setplparam (plotter_params, "BITMAPSIZE", (void *)optarg);
	  break;
	case 'P' << 8:		/* Page size */
	  pl_setplparam (plotter_params, "PAGESIZE", (void *)optarg);
	  break;
	case 'r' << 8:		/* Rotation angle */
	  pl_setplparam (plotter_params, "ROTATION", (void *)optarg);
	  break;
	case 'b' << 8:		/* Bearings requested */
	  bearings = true;
	  break;
	case 'V' << 8:		/* Version */
	  show_version = true;
	  break;
	case 'f' << 8:		/* Fonts */
	  show_fonts = true;
	  break;
	case 'l' << 8:		/* Fonts */
	  do_list_fonts = true;
	  break;
	case 'J':		/* JIS page */
	  if (sscanf (optarg, "%d", &jis_page) <= 0
	      || (jis_page < 33) || (jis_page > 126))
	    {
	      fprintf (stderr,
		       "%s: the JIS page number is bad (it should be in the range 33..126)\n",
		       progname);
	      errcnt++;
	    }
	  else
	    do_jis_page = true;
	  break;
	case 'j':		/* JIS row */
	  if (sscanf (optarg, "%d", &jis_page) <= 0
	      || (jis_page < 1) || (jis_page > 94))
	    {
	      fprintf (stderr,
		       "%s: the JIS row number is bad (it should be in the range 1..94)\n",
		       progname);
	      errcnt++;
	    }
	  else
	    {
	      jis_page += 32;
	      do_jis_page = true;
	    }
	  break;
	case 'h' << 8:		/* Help */
	  show_usage = true;
	  break;

	default:
	  errcnt++;
	  break;
	}
    }

  if (errcnt > 0)
    {
      fprintf (stderr, "Try `%s --help' for more information\n", progname);
      return EXIT_FAILURE;
    }
  if (show_version)
    {
      display_version (progname, written, copyright);
      return EXIT_SUCCESS;
    }
  if (do_list_fonts)
    {
      int success;

      success = list_fonts (output_format, progname);
      if (success)
	return EXIT_SUCCESS;
      else
	return EXIT_FAILURE;
    }
  if (show_fonts)
    {
      int success;

      success = display_fonts (output_format, progname);
      if (success)
	return EXIT_SUCCESS;
      else
	return EXIT_FAILURE;
    }
  if (show_usage)
    {
      display_usage (progname, hidden_options, usage_appendage, 2);
      return EXIT_SUCCESS;
    }

  if (option_font_name == NULL && optind >= argc)
    {
      fprintf (stderr, "%s: no font or fonts are specified\n", progname);
      return EXIT_FAILURE;
    }

  if (do_jis_page)
    {
      if ((!((option_font_name == NULL && optind == argc - 1)
	    || (option_font_name && optind >= argc)))
	  || (option_font_name && strcasecmp (option_font_name, "HersheyEUC") != 0)
	  || (!option_font_name && strcasecmp (argv[optind], "HersheyEUC") != 0))
	{
	  fprintf (stderr, "%s: a JIS page can only be specified for the HersheyEUC font\n", progname);
	  return EXIT_FAILURE;
	}	  
    }

  if ((plotter = pl_newpl_r (output_format, NULL, stdout, stderr, 
			     plotter_params)) == NULL)
    {
      fprintf (stderr, "%s: error: the plot device could not be created\n", progname);
      return EXIT_FAILURE;
    }

  if (option_font_name)
    /* user specifed a font with -F */
    {
      if (do_font (plotter, option_font_name, upper_half, pen_color, numbering_font_name, title_font_name, bearings, base, jis_page, do_jis_page) == false)
	return EXIT_FAILURE;
    }
  
  if (optind < argc)
    /* 1 or more fonts named explicitly on command line */
    {
      for (; optind < argc; optind++)
	{
	  char *font_name;

	  font_name = argv[optind];
	  if (do_font (plotter, font_name, upper_half, pen_color, numbering_font_name, title_font_name, bearings, base, jis_page, do_jis_page) == false)
	    return EXIT_FAILURE;
	}
    }

  /* clean up */
  retval = EXIT_SUCCESS;
  if (pl_deletepl_r (plotter) < 0)
    {
      fprintf (stderr, "%s: error: the plot device could not be deleted\n", progname);
      retval = EXIT_FAILURE;
    }
  pl_deleteplparams (plotter_params);

  return retval;
}

#define NUM_ROWS 12

#define SIZE 1.0
#define MAX_TITLE_LENGTH 0.9

#define HSPACING 0.1	/* must have 8*HSPACING < SIZE */
#define VSPACING 0.06	/* must have 12*VSPACING < SIZE-LINE_VOFFSET */

#define LINE_HOFFSET (0.5 * ((SIZE) - 8 * (HSPACING)))
#define LINE_VOFFSET 0.15	/* top line down from top of display by this */
#define CHAR_HOFFSET ((LINE_HOFFSET) + 0.5 * (HSPACING))
#define CHAR_VOFFSET ((LINE_VOFFSET) + 0.5 * (VSPACING))

#define TOP (SIZE - (LINE_VOFFSET))
#define BOTTOM ((SIZE - (LINE_VOFFSET)) - 12 * (VSPACING))
#define LEFT (LINE_HOFFSET)
#define RIGHT (LINE_HOFFSET + 8 * HSPACING)

#define FONT_SIZE 0.04
#define TITLE_FONT_SIZE 0.045
#define NUMBERING_FONT_SIZE 0.015

/* shifts of numbers in grid cells leftward and upward */
#define N_X_SHIFT 0.015
#define N_Y_SHIFT 0.05

bool
do_font (plPlotter *plotter, const char *name, bool upper_half, char *pen_color_name, char *numbering_font_name, char *title_font_name, bool bearings, enum radix base, int jis_page, bool do_jis_page)
{
  char buf[16];
  char numbuf[16];
  char suffixbuf[16];
  char *titlebuf;
  const char *suffix;
  double title_width;
  int i, j, bottom_octet, top_octet;

  if (do_jis_page)
    {
      switch (base)
	{
	case DECIMAL:
	default:
	  sprintf (suffixbuf, " (row %d)", jis_page - 32);
	  break;
	case OCTAL:
	  sprintf (suffixbuf, " (row 0%o)", jis_page - 32);
	  break;
	case HEXADECIMAL:
	  sprintf (suffixbuf, " (row 0x%X)", jis_page - 32);
	  break;
	}
      suffix = suffixbuf;
    }
  else
    suffix = upper_half ? " (upper half)" : " (lower half)";
  titlebuf = (char *)xmalloc (strlen (name) + strlen (suffix) + 1);
  strcpy (titlebuf, name);
  strcat (titlebuf, suffix);

  if (pl_openpl_r (plotter) < 0)
    {
      fprintf (stderr, "%s: error: the plot device could not be opened\n", progname);
      return false;
    }

  pl_fspace_r (plotter, 0.0, 0.0, (double)SIZE, (double)SIZE);
  pl_erase_r (plotter);
  if (pen_color_name)
    pl_pencolorname_r (plotter, pen_color_name);
  
  pl_fmove_r (plotter, 0.5 * SIZE, 0.5 * (SIZE + TOP));
  if (title_font_name)
    pl_fontname_r (plotter, title_font_name);
  else
    pl_fontname_r (plotter, name);
  pl_ffontsize_r (plotter, (double)(TITLE_FONT_SIZE));

  title_width = pl_flabelwidth_r (plotter, titlebuf);
  if (title_width > MAX_TITLE_LENGTH)
    /* squeeze title to fit */
    pl_ffontsize_r (plotter, 
		    (double)(TITLE_FONT_SIZE) * (MAX_TITLE_LENGTH / title_width));

  /* print title */
  pl_alabel_r (plotter, 'c', 'c', titlebuf);

  if (do_jis_page)
    bottom_octet = 4;
  else			/* ordinary map */
    {
      if (upper_half)
	bottom_octet = 20;
      else
	bottom_octet = 4;
    }
  top_octet = bottom_octet + NUM_ROWS - 1;

  /* draw grid */

  pl_linemod_r (plotter, "solid");
  pl_fbox_r (plotter, LEFT, BOTTOM, RIGHT, TOP);
  for (i = 1; i <= 7; i++)
    /* boustrophedon */
    {
      if (i % 2)
	pl_fline_r (plotter, 
		    LINE_HOFFSET + i * HSPACING, BOTTOM, 
		    LINE_HOFFSET + i * HSPACING, TOP);
      else
	pl_fline_r (plotter, 
		    LINE_HOFFSET + i * HSPACING, TOP,
		    LINE_HOFFSET + i * HSPACING, BOTTOM);
    }      
  for (j = 1; j <= 11; j++)
    /* boustrophedon */
    {
      if (j % 2)
	pl_fline_r (plotter, 
		    RIGHT, TOP - j * VSPACING,
		    LEFT, TOP - j * VSPACING);
      else
	pl_fline_r (plotter,  
		    LEFT, TOP - j * VSPACING,
		    RIGHT, TOP - j * VSPACING);
    }

  /* number grid cells */

  if (numbering_font_name)
    pl_fontname_r (plotter, numbering_font_name);
  else				/* select default font */
    pl_fontname_r (plotter, "");
  pl_ffontsize_r (plotter, (double)(NUMBERING_FONT_SIZE));
  if (bearings)
    pl_linemod_r (plotter, "dotted");
  for (i = bottom_octet; i <= top_octet; i++)
    for (j = 0; j < 8; j++)
      {
	int row, column, charnum;

	row = i - bottom_octet;	/* row, 0..11 */
	column = j;		/* column, 0..7 */

	charnum = (unsigned char)(8 * i + j);
	if (charnum == 127)	/* 0xff not a legitimate glyph */
	  continue;
	if (do_jis_page && charnum == 32)
	  continue;		/* 0x20 not legitimate for JIS */

	switch (base)
	  {
	  case HEXADECIMAL:
	    write_two_bytes (charnum - (do_jis_page ? 32 : 0), numbuf, 16);
	    break;
	  case DECIMAL:
	  default:
	    write_three_bytes (charnum - (do_jis_page ? 32 : 0), numbuf, 10);
	    break;
	  case OCTAL:
	    write_three_bytes (charnum - (do_jis_page ? 32 : 0), numbuf, 8);
	    break;
	  }
	
	pl_fmove_r (plotter,
		    (double)(LINE_HOFFSET + HSPACING * (column + 1 - N_X_SHIFT)),
		    (double)(SIZE - (LINE_VOFFSET + VSPACING * (row + 1 - N_Y_SHIFT))));
	pl_alabel_r (plotter, 'r', 'x', numbuf);
      }

  /* fill grid cells with characters */

  pl_fontname_r (plotter, name);
  pl_ffontsize_r (plotter, (double)(FONT_SIZE));
  for (i = bottom_octet; i <= top_octet; i++)
    for (j = 0; j < 8; j++)
      {
	int row, column, charnum;

	row = i - bottom_octet;	/* row, 0..11 */
	column = j;		/* column, 0..7 */
	
	charnum = (unsigned char)(8 * i + j);
	if (charnum == 127)	/* 0xff not a legitimate glyph */
	  continue;
	if (do_jis_page && charnum == 32)
	  continue;		/* 0x20 not legitimate for JIS */

	if (!do_jis_page)
	  {
	    buf[0] = charnum;
	    buf[1] = '\0';
	  }
	else			/* do JIS page */
	  {
	    /* two bytes, set high bits on both page and character */
	    buf[0] = jis_page + 0x80;
	    buf[1] = charnum + 0x80;
	    buf[2] = '\0';
	  }

	pl_fmove_r (plotter, 
		    (double)(LINE_HOFFSET + HSPACING * (column + 0.5)),
		    (double)(SIZE - (LINE_VOFFSET + VSPACING * (row + 0.5))));
	/* place glyph on page */
	pl_alabel_r (plotter, 'c', 'c', (char *)buf);
	if (bearings)
	  {
	    double halfwidth;

	    /* compute glyph width */
	    halfwidth = 0.5 * pl_flabelwidth_r (plotter, (char *)buf);
	    if (halfwidth == 0.0)
	      /* empty glyph, draw only one vertical dotted line */
	      pl_fline_r (plotter,
			  (double)(CHAR_HOFFSET + HSPACING * column),
			  (double)(SIZE - (CHAR_VOFFSET + VSPACING * (row - 0.5))),
			  (double)(CHAR_HOFFSET + HSPACING * column),
			  (double)(SIZE - (CHAR_VOFFSET + VSPACING * (row + 0.5))));
	    else
	      /* draw vertical dotted lines to either side of glyph */
	      {
		pl_fline_r (plotter, 
			    (double)(CHAR_HOFFSET + HSPACING * column - halfwidth),
			    (double)(SIZE - (CHAR_VOFFSET + VSPACING * (row - 0.5))),
			    (double)(CHAR_HOFFSET + HSPACING * column - halfwidth),
			    (double)(SIZE - (CHAR_VOFFSET + VSPACING * (row + 0.5))));
		pl_fline_r (plotter, 
			    (double)(CHAR_HOFFSET + HSPACING * column + halfwidth),
			    (double)(SIZE - (CHAR_VOFFSET + VSPACING * (row - 0.5))),
			    (double)(CHAR_HOFFSET + HSPACING * column + halfwidth),
			    (double)(SIZE - (CHAR_VOFFSET + VSPACING * (row + 0.5))));
	      }
	  }
      }
  
  if (pl_closepl_r (plotter) < 0)
    {
      fprintf (stderr, "%s: error: the plot device could not be closed\n", progname);
      return false;
    }
  return true;
}

/* Write an integer that is < (radix)**4 as three ascii bytes with respect
   to the specified radix.  Initial zeroes are converted to spaces. */
void 
write_three_bytes (int charnum, char *numbuf, int radix)
{
  int i;

  numbuf[0] = charnum / (radix * radix);
  numbuf[1] = (charnum - (radix * radix) * (charnum / (radix * radix))) / radix;
  numbuf[2] = charnum % radix;
  numbuf[3] = '\0';

  for (i = 0 ; i <= 2 ; i++)
    {
      if (numbuf[i] >= 10)
	numbuf[i] += ('A' - 10);
      else 
	numbuf[i] += '0';
    }
  if (numbuf[0] == '0')
    {
      numbuf[0] = ' ';
      if (numbuf[1] == '0')
	numbuf[1] = ' ';
    }
}

/* Write an integer that is < (radix)**3 as two ascii bytes with respect to
   the specified radix.  Initial zeroes are converted are not changed. */
void 
write_two_bytes (int charnum, char *numbuf, int radix)
{
  int i;

  numbuf[0] = charnum / radix;
  numbuf[1] = charnum % radix;
  numbuf[2] = '\0';

  for (i = 0 ; i <= 1 ; i++)
    {
      if (numbuf[i] >= 10)
	numbuf[i] += ('A' - 10);
      else 
	numbuf[i] += '0';
    }
}
